File: packages/f#/Runner.fs
namespace FSharpCgdk

open System.Collections.Generic
open FSharpCgdk.Model

module Runner = 
    let emptyAction() = {
        target_velocity_x = 0.0
        target_velocity_y = 0.0
        target_velocity_z = 0.0
        jump_speed = 0.0
        use_nitro = false }


    let startRunner (rpc : RemoteProcessClient) (token : string) = 
        rpc.writeToken token

        let rules = rpc.readRules() |> Option.get
        let mutable gameOpt = rpc.readGame() 
        let mutable actions = Map.empty

        while gameOpt <> None do 
            actions <- Map.empty
            let game = Option.get gameOpt
            let teammates = game.robots |> Array.filter (fun x -> x.is_teammate)        
            for robot in teammates do 
                let action = emptyAction()
                MyStrategy.act(robot, rules, game, action)
                actions <- Map.add (string robot.id) action actions
            rpc.write (actions, MyStrategy.customRendering())
            gameOpt <- rpc.readGame()

    
    let templateArgs = "127.0.0.1", "31001", "0000000000000000"
    
    [<EntryPoint>]
    let main(args : string array) = 
        let host, port, token = 
            match args with
            | [| host; port; token |] -> host, port, token
            | otherwise -> templateArgs
        let rpc = RemoteProcessClient(host, int port)
        startRunner rpc token   
        0  

File: packages/c++17/Runner.cpp
#include <memory>

#include "Runner.h"
#include "MyStrategy.h"

using namespace model;
using namespace std;

int main(int argc, char* argv[]) {
    if (argc == 4) {
        Runner runner(argv[1], argv[2], argv[3]);
        runner.run();
    } else {
        Runner runner("127.0.0.1", "31001", "0000000000000000");
        runner.run();
    }

    return 0;
}

Runner::Runner(const char* host, const char* port, const char* token)
    : remoteProcessClient(host, atoi(port)), token(token) {
}

void Runner::run() {
    unique_ptr<Strategy> strategy(new MyStrategy);
    unique_ptr<Game> game;
    unordered_map<int, Action> actions;
    remoteProcessClient.write_token(token);
    unique_ptr<Rules> rules = remoteProcessClient.read_rules();
    while ((game = remoteProcessClient.read_game()) != nullptr) {
        actions.clear();
        for (const Robot& robot : game->robots) {
            if (robot.is_teammate) {
                strategy->act(robot, *rules, *game, actions[robot.id]);
            }
        }
        remoteProcessClient.write(actions, strategy->custom_rendering());
    }
}
